### PR TITLE
Add MatrixFreeTools::ElementBirthAndDeathMatrixFree

### PR DESCRIPTION
**Pull request by peterrum**
references https://github.com/dealii/dealii/issues/12102

~depends on #14122~

### CODE REVIEW

**Comment by peterrum** [comment]
@sebproell and I have been working on resolving issue https://github.com/dealii/dealii/issues/12102. We propose not to integrate the notion of `FENothing` in `MatrixFree` but instead have the logic inside of wrapper with a `MatrixFree`-like interface (`loop` and `cell_loop`). After some generalizations and bug fixes, which all have been merged by now, this is possible. What do people think?

FYI @Rombur you have open the original issue. Would this approach work for you?

File: include/deal.II/matrix_free/tools.h [hunk 1]
```diff
@@ -157,6 +157,179 @@ namespace MatrixFreeTools
...
+   * in which context valid cells are alive and invalid cells are dead.
+   */
+  template <int dim,
+            typename Number,
+            typename VectorizedArrayType = VectorizedArray<Number>>
+  class ElementBirthAndDeathMatrixFree
+  {
+  public:
+    /**
+     * Strcut that helps to configure ElementBirthAndDeathMatrixFree.
```
Comment (sebproell) [inline include/deal.II/matrix_free/tools.h:179]:
```suggestion
     * Struct that helps to configure ElementBirthAndDeathMatrixFree.
```

Comment (Rombur) [inline include/deal.II/matrix_free/tools.h:179]:
```suggestion
     * Struct that helps to configure ElementBirthAndDeathMatrixFree.
```

File: include/deal.II/matrix_free/tools.h [hunk 2]
```diff
@@ -157,6 +157,179 @@ namespace MatrixFreeTools
...
 
 
+
+  /**
+   * A wrapper around MatrixFree to help users to deal with DoFHandlers,
+   * where there are cells without degrees of freedom, i.e., with
+   * FENothing assigned to them. In the following we call such cells
+   * invalid. All other cells are valid. In contrast to MatrixFree,
+   * this class skips invalid cells and faces between valid and
+   * invalid cells are treated as boundary faces.
```
Comment (sebproell) [inline include/deal.II/matrix_free/tools.h:167]: Maybe mention how to figure out if a face is an actual boundary face or a valid-invalid boundary face inside the user-supplied boundary operation?

File: include/deal.II/matrix_free/tools.h [hunk 3]
```diff
@@ -157,6 +157,179 @@ namespace MatrixFreeTools
...
+        [&](const auto &matrix_free,
+            auto &      dst,
+            const auto &src,
+            const auto  range) {
+          const auto category = matrix_free.get_face_range_category(range);
+
+          const unsigned int type =
+            static_cast<unsigned int>(category.first == fe_index_valid) +
+            static_cast<unsigned int>(category.second == fe_index_valid);
+
```
Comment (sebproell) [inline include/deal.II/matrix_free/tools.h:303]: Would `Assert(type != 0, ExcInternalError())` be useful here?

File: include/deal.II/matrix_free/tools.h [hunk 4]
```diff
@@ -157,6 +157,179 @@ namespace MatrixFreeTools
...
+        cell_operation(matrix_free, dst, src, range);
+      };
+
+      matrix_free->template cell_loop<VectorTypeOut, VectorTypeIn>(
+        ebd_cell_operation, dst, src, zero_dst_vector);
+    }
+
+    /**
+     * Loop over all valid cells and faces. Faces between valid and
+     * invalid cells are treated as boundary faces.
```
Comment (Rombur) [inline include/deal.II/matrix_free/tools.h:255]: I agree that it's what most people want but it is very different than what we do everywhere else. That's fine with me but that's something we need to think about. Since the face is treated as a boundary, which boundary ID is used?

**Review by sebproell — COMMENTED** [review]
Thanks @peterrum!

**Review by Rombur — COMMENTED** [review]
This looks good to me. My main concern is that we introduce the concept of boundary inside the domain which we have always forbidden.

File: include/deal.II/matrix_free/tools.h (hunk 3 above)
Comment (peterrum) [reply include/deal.II/matrix_free/tools.h:304]: That is invalid face and nothing is done and the face is skipped. I have added a comment.

File: include/deal.II/matrix_free/tools.h (hunk 2 above)
Comment (peterrum) [reply include/deal.II/matrix_free/tools.h:167]: I would use this function: https://github.com/dealii/dealii/pull/14122 :)

File: include/deal.II/matrix_free/tools.h (hunk 4 above)
Comment (peterrum) [reply include/deal.II/matrix_free/tools.h]: > I agree that it's what most people want but it is very different than what we do everywhere else. That's fine with me but that's something we need to think about. 

That is why it is not part of `MatrixFree` actual. This is helper class tailored for such purposes.

> Since the face is treated as a boundary, which boundary ID is used?

I would say: `numbers::internal_face_boundary_id`. See https://github.com/dealii/dealii/pull/14122

File: doc/news/changes/minor/20220709Munch [hunk 5]
```diff
@@ -0,0 +1,7 @@
+New: The new class MatrixFreeTools::ElementBirthAndDeathMatrixFree
+is a wrapper around MatrixFree, which helps users to deal with 
+DoFHandlers, where there are cells without degrees of freedom, 
+i.e., with FENothing assigned to them. The class helps to implement 
+"element birth and death technique".
```
Comment (tjhei) [inline doc/news/changes/minor/20220709Munch:5]: Is this a term used anywhere else on the literature? If not, then I would suggest to with something like activation/ deactivation instead.

File: include/deal.II/matrix_free/tools.h [hunk 6]
```diff
@@ -157,6 +157,183 @@ namespace MatrixFreeTools
...
+
+      const auto &fe_collection =
+        matrix_free.get_dof_handler(additional_data.dof_index)
+          .get_fe_collection();
+
+      for (unsigned int i = 0; i < fe_collection.size(); ++i)
+        if (fe_collection[i].n_dofs_per_cell() > 0)
+          valid_fe_indices.insert(i);
+
+      AssertDimension(valid_fe_indices.size(), 1);
```
Comment (bergbauer) [inline include/deal.II/matrix_free/tools.h:216]: Can we allow more than one `valid_fe_indices`? Or create an `invalid_fe_index` and check for it later?

Comment (bergbauer) [reply include/deal.II/matrix_free/tools.h:216]: > Also, I guess what you need in your case is to not loop over faces neighboring FE_Nothing.

Yes, this should only be optional from my perspective. For me this class can definitely be useful, currently I skip FENothing cells (or cell batches) manually.

File: doc/news/changes/minor/20220709Munch (hunk 5 above)
Comment (tjhei) [reply doc/news/changes/minor/20220709Munch:5]: Yes, something like that sounds better to me. Would that be acceptable?

Comment (peterrum) [reply doc/news/changes/minor/20220709Munch:5]: Renamed!

**Comment by peterrum** [comment]
With https://github.com/dealii/dealii/pull/14122 being merged, I have rebased this PR! Ready from my side.

File: doc/news/changes/minor/20220709Munch [hunk 7]
```diff
@@ -0,0 +1,7 @@
+New: The new class MatrixFreeTools::ElementActivationAndDeactivationMatrixFree
+is a wrapper around MatrixFree, which helps users to deal with 
+DoFHandlers, where there are cells without degrees of freedom, 
+i.e., with FENothing assigned to them. The class helps to implement
```
Comment (kronbichler) [inline doc/news/changes/minor/20220709Munch:4]: There are too many sub-clauses in this description. How about
```suggestion
is a wrapper around MatrixFree designed to deal with 
DoFHandler objects involving cells without degrees of freedom, i.e.,
cells using FENothing as element type. The class helps to implement the
```

File: include/deal.II/matrix_free/tools.h [hunk 8]
```diff
@@ -157,6 +157,184 @@ namespace MatrixFreeTools
     const unsigned int first_selected_component = 0);
 
 
+
+  /**
+   * A wrapper around MatrixFree to help users to deal with DoFHandlers,
+   * where there are cells without degrees of freedom, i.e., with
+   * FENothing assigned to them. In the following we call such cells
```
Comment (kronbichler) [inline include/deal.II/matrix_free/tools.h:164]: As above, this is written in a way that makes it hard to follow.

File: include/deal.II/matrix_free/tools.h [hunk 9]
```diff
@@ -157,6 +157,184 @@ namespace MatrixFreeTools
...
+   * deactivated. All other cells are activated. In contrast to MatrixFree,
+   * this class skips deactivated cells and faces between activated and
+   * deactivated cells are treated as boundary faces.
+   *
+   * The class helps to implement "element birth and death technique",
+   * in which context activated cells are alive and deactivated cells are dead.
```
Comment (kronbichler) [inline include/deal.II/matrix_free/tools.h:170]: Do you mean
```suggestion
   * in whose context activated cells are alive and deactivated cells are dead.
```
In any case, this does not really say much about what "element birth and death technique" actually is, you just repeat the term in other words (deactivated/activated).

File: include/deal.II/matrix_free/tools.h [hunk 10]
```diff
@@ -157,6 +157,184 @@ namespace MatrixFreeTools
...
 
 
+
+  /**
+   * A wrapper around MatrixFree to help users to deal with DoFHandlers,
+   * where there are cells without degrees of freedom, i.e., with
+   * FENothing assigned to them. In the following we call such cells
+   * deactivated. All other cells are activated. In contrast to MatrixFree,
+   * this class skips deactivated cells and faces between activated and
+   * deactivated cells are treated as boundary faces.
```
Comment (kronbichler) [inline include/deal.II/matrix_free/tools.h:167]: I think it would be important to say that this class considers only the first `FE` in an `hp::FECollection` that has non-zero number of unknowns (if I read the code correctly).

File: include/deal.II/matrix_free/tools.h [hunk 11]
```diff
@@ -157,6 +157,184 @@ namespace MatrixFreeTools
...
+
+      std::set<unsigned int> valid_fe_indices;
+
+      const auto &fe_collection =
+        matrix_free.get_dof_handler(additional_data.dof_index)
+          .get_fe_collection();
+
+      for (unsigned int i = 0; i < fe_collection.size(); ++i)
+        if (fe_collection[i].n_dofs_per_cell() > 0)
+          valid_fe_indices.insert(i);
```
Comment (kronbichler) [inline include/deal.II/matrix_free/tools.h:215]: Why do you use a `set` here? I see no reason to use that data structure, as you simply collect the indices that are assigned. If you want a single index, a simple integer would suffice (that you check for `!= numbers::invalid_unsigned_int`); if you want more, I would use a vector.

**Review by kronbichler — APPROVED** [review]

File: include/deal.II/matrix_free/tools.h (hunk 10 above)
Comment (peterrum) [reply include/deal.II/matrix_free/tools.h:167]: I have added a comment. To lift this restriction is the next step; see https://github.com/dealii/dealii/pull/14121#discussion_r919035525.

**Comment by peterrum** [comment]
@kronbichler I have incorporated your suggestion! Thanks.

**Review by kronbichler — APPROVED** [review]
